2단계 리뷰 요청드립니다.

itdar opened the pull request:
안녕하세요 동철님! 2단계 리뷰 요청 드립니다.
생각보다 양도 많고 어렵고 끝이 안보여서 고생했네요 ㅎㅎ;;
tdd로 구현하면서 리팩토링을 같이 진행했지만 아직 할게 많이 남았습니다.. 리뷰가 너무 늦어지면 안될 것 같아서 일단 요청 보내는데 양해 부탁드립니다..ㅠㅠ 많은 첨삭 부탁드려요..

현재 구조는 핸들러 -> 디스패처서블릿(요청과 응답을 전달) -> 서블릿(컨트롤러) 형태로 되어있습니다.

요청의 종류 (HttpMethod, Path, Resource 등) 에 따라서 분기하는게 어떻게 해야 깔끔할지 아이디어가 떠오르지 않아서 억지로 구현해놓은 느낌입니다.. 좋은 방법이 있다면 힌트 부탁드립니다..ㅎㅎ;

감사합니다!!

itdar committed:
doc: 요구사항 내용 추가

itdar committed:
feat: Step 1 구현

itdar committed:
refactor: Request 파싱 관련 리팩토링

- HttpRequest -> RequestLine(StartLine) 구조로 변경
- RequestProtocol enum 추가 (프로토콜의 종류, 버전 통합 관리)

itdar committed:
refactor: 동철님 리뷰사항 수정 커밋 #1

- 매직넘버 제거
- Collections.emptyMap() 적용
- 빠진 테스트 추가 및 테스트 의미에 맞게 설명 수정

itdar committed:
docs: 2단계 요구사항 정리

itdar committed:
chore: 패키지 정리

itdar committed:
refactor: 프로토콜의 매직넘버, 매직리터럴 상수화

itdar committed:
Merge branch 'itdar' into step2

itdar committed:
feat: 기능요구사항 1 구현

- http://localhost:8080/index.html 로 접속했을 때 webapp 디렉토리의 index.html 파일을 읽어 클라이언트에 응답한다.

itdar committed:
test: 빠진 테스트케이스 추가

- Path, QueryStrings 관련 테스트 추가

itdar committed:
feat: 요구사항 2 추가

- “회원가입” 메뉴를 클릭하면 http://localhost:8080/user/form.html 으로 이동
- HTML과 URL을 비교해 보고 사용자가 입력한 값을 파싱해 model.User 클래스에 저장

itdar committed:
feat: 요구사항 2 추가

- “회원가입” 메뉴를 클릭하면 http://localhost:8080/user/form.html 으로 이동
- HTML과 URL을 비교해 보고 사용자가 입력한 값을 파싱해 model.User 클래스에 저장

itdar committed:
refactor: 기능요구사항 2 리팩토링

- 서버에서 지원해주는 api / template(page) 확인 추가
- 응답메서드는 응답클래스 만들기 전 정적메서드로 사용하도록 추출
- 예상되는 예외에 대하여 BadRequest 반환 (클라이언트에서 유효하지 않은 쿼리스트링 들어오는 것 등)

itdar committed:
Merge branch 'itdar' into step2

itdar committed:
refactor: 기능요구사항 2 이후 리팩토링

- 디스패처서블릿 추가 (요청핸들러 내에서 요청 분기점 역할)
- HttpResponse, HttpBody, HttpHeader 추가
- StatusCode 추가
- Support 클래스 추가 (서버에서 지원해주는 요청의 경로인지 확인)

itdar committed:
feat: 기능 요구사항 3 구현

- form.html (회원가입 GET -> POST)
- PathAndQueryStrings / QueryStrings 클래스 분리
- UserCreateServlet test 추가
- HttpRequest 내에 Header 클래스 추가

itdar committed:
feat: 기능 요구사항 4 구현

- 회원가입 후 index.html 페이지로 redirection (302, Found)

itdar committed:
feat: 기능 요구사항 5 구현

- 로그인 성공/실패 후 쿠키저장 및 페이지이동

itdar committed:
feat: 기능요구사항 6 구현

- 로그인 한 유저가 전체회원 목록 조회 가능 (java handlebars 사용)

itdar committed:
feat: 기능 요구사항 7 구현

- static 리소스 파일 요청의 경우 요청 확장자에 맞는 컨텐츠타입을 넣어 바디의 바이트 함께 반환

itdar committed:
Merge branch 'itdar' into step2

# Conflicts:
#	README.md
#	src/main/java/exception/ExceptionStrings.java
#	src/test/java/webserver/HttpRequestTest.java
#	src/test/java/webserver/enums/RequestProtocolTest.java

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -52,4 +87,36 @@ private void responseBody(DataOutputStream dos, byte[] body) {
+            String line;
+            line = br.readLine();
+            if (line == null) {
+                return Optional.empty();
```
Exception을 던지지 않은 이유가 있을까요 ?

itdar replied:
엇 그러네요 ㅎㅎ; 요청 첫줄이 없으면 예외를 던져보겠습니다!

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -52,4 +87,36 @@ private void responseBody(DataOutputStream dos, byte[] body) {
         }
     }
+
+    private Optional<HttpRequest> parseRequest(InputStream is) {
```
HttpRequest의 생성 로직이 ReuqestHandler 클래스에 존재하는것이 어색해보이네요. 
HttpRequest의 생성자를 활용해보는건 어떨까요 ?

itdar replied:
오 옙 감사합니다
HttpRequest 안에서 파싱해서 만들면서 아래 짚어주신 HttpHeader 까지 생성하도록 수정해보겠습니다!~

itdar replied:
수정하다보니까 당시에 inputStream 을 Request 객체 안으로 넣는게 어색해서 그렇게 했던 것 같네요 ㅎㅎ;
수정해보았습니다!!

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -52,4 +87,36 @@ private void responseBody(DataOutputStream dos, byte[] body) {
+                line = br.readLine();
+                httpRequest.addHeader(line);
+                logger.debug(line);
+            }
```
Header 생성 로직도 분리가 가능할것 같네요 😄

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -19,31 +29,56 @@ public RequestHandler(Socket connectionSocket) {
+                .orElseThrow(IllegalArgumentException::new);
+            HttpResponse httpResponse = new HttpResponse();
+
+            DispatcherServlet.INSTANCE.serve(httpRequest, httpResponse);
```
DispatcherServlet 생성 좋네요 ㅎㅎ 👍

itdar replied:
아는 이름이라서 써보았습니다.. 👍🏼

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -19,31 +29,56 @@ public RequestHandler(Socket connectionSocket) {
-            dos.writeBytes("Content-Type: text/html;charset=utf-8\r\n");
-            dos.writeBytes("Content-Length: " + lengthOfBodyContent + "\r\n");
-            dos.writeBytes("\r\n");
+            responseStartLine(dos, httpResponse.responseLine());
```
Http Response의 첫번째 라인은 Status Line으로 정의하곤 합니다 😄 

https://developer.mozilla.org/ko/docs/Web/HTTP/Messages#http_%EC%9D%91%EB%8B%B5

itdar replied:
오 감사합니다!! Request 는 Start Line, Response 는 Status Line.. 👍🏼

sah3122 commented on src/main/java/webserver/domain/ContentType.java:
```diff
@@ -0,0 +1,29 @@
+package webserver.domain;
+
+public enum ContentType {
```
Content Type 정의가 쉽지 않았을텐데 정리 잘해주셨네요 👍

itdar replied:
감사합니다~~! MDN 에서 확장자랑 대응되는 표를 제공해주는 것과 이것저것 참고했습니다!!

sah3122 commented on src/main/java/webserver/enums/StatusCode.java:
```diff
@@ -0,0 +1,25 @@
+package webserver.enums;
+
+public enum StatusCode {
```
Status 정의 👍

sah3122 commented on src/main/java/webserver/request/HttpBody.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Objects;
+import webserver.utils.QueryStrings;
+
+public class HttpBody {
+
+    private final QueryStrings queryStrings;
```
HttpBody는 QueryStrings가 아닌 Key, Value 구조를 가지는 상태값을 선언하는게 어떨까요 ㅎㅎ 

추가로 Http Response Body , Request Body를 구분하는게 의미를 확실하게 나타낼수 있을것 같아요.

itdar replied:
옙옙 파싱 시에만 사용하고 key / value 구조로 수정해보겠습니다!
 추가로 요청 / 응답 바디도 분리해보겠습니다~!

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,62 @@
+        this.protocol = Protocol.HTTP_1_1;
+    }
+
+    public void statusOk() {
+        this.statusCode = StatusCode.OK;
+    }
+
+    public void statusFound() {
+        this.statusCode = StatusCode.FOUND;
+    }
+
+    public void statusBadRequest() {
+        this.statusCode = StatusCode.BAD_REQUEST;
+    }
```
반환되는 Status Code 를 모두 정의할 필요가 있을까요 ? 
response.ok 와 같은 구조로 표현할수 있을것 같은데, 고민해보세요 😄

itdar replied:
옙 만들면서도 찝찝했는데 고민을 안해봤네요 ㅎㅎ;;
고민해서 수정해보겠습니다!~

sah3122 replied:
코멘트가 많이 모호했던것 같아요. 
제가 생각하는 구조는 HttpResponse 의 statusCode 만 변경할수 있는 메서드를 제공하는것과, Response Header 와 같은 상태값들중 Status Code 에 따라 정의되어야 하는 값들이 존재할거에요. 
ok, badRequest 와 같은 메소드를 제공한다면 그러한 값들도 추가하는것도 좋은 방법이 될것 같습니다 ㅎㅎ

itdar replied:
아아 옙 설명 감사합니다!!
단순히 상태만 바꾸는 ok 외에 Location 이 필요한 302 Found 같은 것은 추가로 변경하도록 수정해보겠습니다!

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,62 @@
+        httpHeader.putHeader(key, value);
+    }
+
+    public String responseLine() {
```
StatusLine 클래스를 생성하여 관리해보세요 😄

itdar replied:
오 이건 위에 말씀해주신 내용들 수정하면서 같이 수정해보았었습니다!

sah3122 commented on src/main/java/webserver/response/HttpStatusResponse.java:
```diff
@@ -0,0 +1,36 @@
+public class HttpStatusResponse {
+    private static final Logger logger = LoggerFactory.getLogger(HttpStatusResponse.class);
+
+    public static void responseBodRequest400(DataOutputStream dos) {
```
400 에러 정의 좋네요 👍 
한가지 아쉬운건 응답값을 생성하는 코드가 분산되어있는것 같아요. 
HttpResponse 객체에게 메시지를 던질수 있을것 같은데, 도전해보세요 😄

itdar replied:
응닶에 대한 값들 중 나눠져야 할 것 외에는 합쳐서 메시지를 던지는 식으로 수정해보겠습니다!!

sah3122 commented on src/main/java/webserver/servlet/Servlet.java:
```diff
@@ -0,0 +1,10 @@
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+public interface Servlet {
```
Servlet 인터페이스 정의 👍 
@FunctionalInterface 로 선언해도 좋을것 같네요 ㅎㅎ

itdar replied:
엇 그러네요 감사합니다!! serve() 한개로 끝날 것 같지는 않아서 펑셔널인터페이스라는 생각을 아예 못했네요 ㅎㅎ;;
감사합니다!

sah3122 commented on src/main/java/webserver/servlet/UserListServlet.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public void serve(HttpRequest httpRequest, HttpResponse httpResponse) {
+        String cookie = httpRequest.getHeader("Cookie");
```
Cookie 클래스를 만들어 관리해보는건 어떨까요 ?

itdar replied:
옙 감사합니다! 그러면서 쿠키 여러개를 받을 수 있게끔 구현해봐야겠습니다!~

sah3122 commented on src/main/java/webserver/servlet/UserListServlet.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private void handleLoginUser(final HttpResponse httpResponse) {
+        TemplateLoader loader = new ClassPathTemplateLoader();
+        loader.setPrefix("/templates");
+        loader.setSuffix(".html");
+        Handlebars handlebars = new Handlebars(loader);
+
+        try {
+            Template template = handlebars.compile("user/list");
+            String profilePage = template.apply(Collections.singletonMap("users", DataBase.findAll()));
+            byte[] body = profilePage.getBytes(StandardCharsets.UTF_8);
+
+            httpResponse.statusOk();
+            httpResponse.addHeader("Content-Length", Integer.toString(body.length));
+            httpResponse.setBody(body);
+        } catch (IOException e) {
+            throw new IllegalStateException("Handlebars template error");
+        }
```
Template 생성 코드는 재사용 가능하도록 Util 클래스로 분리해보면 어떨까요 ?

itdar replied:
엇 그러네요 Util 클래스로 수정해보겠습니다!

sah3122 commented on src/main/java/webserver/servlet/UserLoginServlet.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void serve(HttpRequest httpRequest, HttpResponse httpResponse) {
+        Map<String, String> bodyQueryStrings = httpRequest.bodyQueryString();
+        String userId = bodyQueryStrings.get("userId");
+        String password = bodyQueryStrings.get("password");
```
```suggestion
        String userId = httpRequest.getQueryString("userId");
        String password = httpRequest.getQueryString("password");
```
어떤 차이가 있을지 고민해보세요 😄

itdar replied:
엇 외부에서 수정가능하게 열려있었군요.. 수정해보겠읍니다!

sah3122 commented on src/main/java/webserver/supporter/SupportApis.java:
```diff
@@ -0,0 +1,29 @@
+import webserver.servlet.UserListServlet;
+import webserver.servlet.UserLoginServlet;
+
+public class SupportApis {
```
지원하는 Servlet 클래스 정의 👍

itdar replied:
크 갈길이 먼데 간간히 섞인 칭찬이 힘이 되네요 감사합니다 ㅋㅋ;

sah3122 commented on src/main/java/webserver/supporter/SupportResources.java:
```diff
@@ -0,0 +1,51 @@
+    private SupportResources() { }
+
+    private static Set<String> resourceMap;
+    static {
```
모든 resource를 정의할수 없을것 같아요. 
지원하는 확장자만 관리할수 있도록 리팩토링 해보세요 😄

itdar replied:
DispatcherServlet 에서
 순수 Api 호출의 경우 먼저 확인해서 처리하고
 리소스인 경우에는, template(html) / static(그 외) 모든 리소스를 처리하게 수정해보았습니다!

리소스 처리 내부에서는 하나로 처리할까 하다가 경로가 다르니 기존의 Template 처리 클래스를 남겨서 분기하여 처리하도록 해두었습니다!

sah3122 commented on src/main/java/webserver/supporter/SupportTemplates.java:
```diff
@@ -0,0 +1,40 @@
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+public final class SupportTemplates {
```
Template 을 관리하기 보단, 리소스 경로가 올바르지 않을 경우 NotFound.html 을 제공하는것이 어떨까요 ?

itdar replied:
오 좋은 것 같습니다! 그렇게 수정해봐야겠습니다

sah3122 commented on src/main/java/webserver/utils/PathAndQueryStrings.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class PathAndQueryStrings {
```
조금더 적절한 네이밍을 고민해보세요. 
Path / QueryStrings 를 분리하는것도 하나의 방법이거나, URI 와 같은  네이밍도 어울릴수 있을것 같아요 😄

itdar replied:
아.. URI 에서 path, query 둘 다 포함할 수 있겠네요! 감사합니다~!

sah3122 commented on src/main/java/webserver/utils/PathAndQueryStrings.java:
```diff
@@ -0,0 +1,57 @@
+    
+
+    private String path;
+    private Map<String, String> queryStringsMap;
```
정의해둔 QueryStrings 클래스를 사용하는것도 좋은 방법이 될거 같아요 😄

itdar replied:
옙 위에 네이밍을 Uri 로 바꾸면서 전체적으로 수정을 해보았습니다!!

Uri (이전에 PathAndQu...) -> Path, QueryStrings

sah3122 commented on src/test/java/db/DataBaseTest.java:
```diff
@@ -0,0 +1,47 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class DataBaseTest {
```
동료 개발자가 테스트 코드만 보고도 어떤 기능을 제공하는지 알수 있도록 DisplayName 으로 테스트의 의도를 확실하게 나타내주세요. 😄

sah3122 replied:
꼼꼼한 테스트 코드 좋습니다 👍

itdar replied:
넵넵.. 좀 더 부지런하게 이름을 잘 달아보겠습니다..!!

sah3122 reviewed (changes requested):
2단계 미션 구현 너무 잘해주셨네요 👍 
사소한 코멘트들과 몇가지 고민해볼만한 코멘트 남겨두었어요. 

미션 진행도중 이해가 힘든 부분이 있다면 편하게 Slack DM 주세요 🙏

itdar committed:
refactor: 동철님 리뷰사항 수정 #1

- 요청/응답 각각 첫줄,헤더,바디 로 구성되도록 수정
  - 각각 요청/응답 에 맞는 클래스명으로 수정
- 요청 파싱부분을 HttpRequest 내부로 이동
- 예외처리 추가
- HttpHeader String 클래스 추가
- 불필요 주석 제거

itdar committed:
refactor: 동철님 리뷰사항 수정 #2

- 함수형인터페이스 어노테이션 추가
- 응답에 대한 수정은 응답객체에 메시지를 던지도록 수정
- 쿠키 클래스 추가
- 핸들바 템플릿 재사용 가능하도록 유틸화
- 외부에서 변경 가능한 콜렉션 필요하지 않다면 닫음
- Resource (static, template) 구조 변경
- 네이밍 수정
- 테스트 상세 이름 추가

sah3122 commented on src/main/java/webserver/domain/Cookies.java:
```diff
@@ -0,0 +1,41 @@
+    private static final String COOKIE_STRING_SPLIT_REGEX = ";";
+    private static final String COOKIE_KEY_VALUE_SPLIT_REGEX = "=";
+
+    private Map<String, String> map;
```
현재 미션에서는 Key - Value 형태의 쿠키가 요구사항을 만족하지만, 더 많은 스펙을 가지고 있으신걸 알고 계실거에요 😄 
혹시 몰라 문서 링크 첨부 합니다 👍 
https://developer.mozilla.org/ko/docs/Web/HTTP/Cookies

itdar replied:
옙 슬랙으로 질문 보내보았습니다!!~

sah3122 commented on src/main/java/webserver/domain/HttpHeader.java:
```diff
@@ -0,0 +1,62 @@
+package webserver.domain;
+
+public class HttpHeader {
```
크 .. 헤더 정리까지 👍

sah3122 commented on src/main/java/webserver/request/HttpRequestBody.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Objects;
+import webserver.utils.QueryStrings;
+
+public class HttpRequestBody {
```
Request Body는, Content-Type에 전달 받는 데이터 타입이 다를수 있을것 같아요. 
이러한 요구사항들은 먼저 인터페이스 분리를 고민해보세요. 😄

itdar replied:
아하.. 옙옙 감사합니다!!

sah3122 reviewed (approved):
꼼꼼한 리뷰 반영 👍 
적지않은 코드가 변경되었는데 너무 잘해주셨습니다 💯 
다음 미션도 기대하겠습니다. 🙏
